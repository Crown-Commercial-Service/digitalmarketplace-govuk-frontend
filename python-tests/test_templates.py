from pathlib import Path
from typing import Iterable
from unittest import mock

import jinja2
import pytest

root_path = Path(__file__, "..", "..").resolve()
src_path = root_path / "src"


def find_njk_templates(path: Path) -> Iterable[Path]:
    """Walk a directory tree looking for njk files"""
    if path.is_dir():
        for child in path.iterdir():
            yield from find_njk_templates(child)
    elif path.is_file():
        if path.suffix == ".njk":
            yield path


digitalmarketplace_templates = tuple(list(find_njk_templates(src_path / "digitalmarketplace")))


@pytest.fixture
def env():
    class GOVUKFrontendJinjaLoader(jinja2.PackageLoader):
        # work around the fact the filenames are different in govuk_frontend_jinja
        def get_source(self, environment, template):
            if template.endswith(".njk"):
                template = template[:-4] + ".html"
            return super().get_source(environment, template)

    loader = jinja2.PrefixLoader({
        "digitalmarketplace": jinja2.FileSystemLoader(src_path / "digitalmarketplace"),
        "govuk": GOVUKFrontendJinjaLoader("govuk_frontend_jinja"),
        "govuk_frontend_jinja": jinja2.PackageLoader("govuk_frontend_jinja"),
    })

    env = jinja2.Environment(loader=loader)
    env.globals["url_for"] = mock.Mock(return_value="#")
    return env


class MockParams(mock.MagicMock):
    def __repr__(self):
        # make sure the str is deterministic
        _name_list = [self._mock_new_name]
        _parent = self._mock_new_parent
        last = self

        dot = "."
        if _name_list == ["()"]:
            dot = ""
        seen = set()
        while _parent is not None:
            last = _parent

            _name_list.append(_parent._mock_new_name + dot)
            dot = "."
            if _parent._mock_new_name == "()":
                dot = ""

            _parent = _parent._mock_new_parent

            if id(_parent) in seen:
                break
            seen.add(id(_parent))

        _name_list = list(reversed(_name_list))
        _first = last._mock_name or "params"
        if len(_name_list) > 1:
            if _name_list[1] not in ("()", "()."):
                _first += "."
        _name_list[0] = _first
        name = "".join(_name_list)

        return f"<MockParam {name}>"

    def _get_child_mock(self, **kw):
        return MockParams(**kw)


@pytest.fixture
def params():
    return MockParams()


@pytest.fixture(params=digitalmarketplace_templates)
def template_path(request):
    template_path = str(request.param.relative_to(src_path))
    return template_path


def test_template_loads(env, template_path):
    env.get_template(template_path)


def test_template_renders(env, params, template_path, snapshot):
    template = env.get_template(template_path)
    template_output = template.render(params=params)
    assert template_output == snapshot


def test_dm_alert_with_no_attributes(env, snapshot):
    """Test that we don't get an undefined error when rendering dmAlert"""
    # TODO: this test is here to catch a regression, ideally it would be
    # replaced with a proper test suite for the component (see
    # https://trello.com/c/gw8q0xdl)
    template = env.from_string("""
        {% from "digitalmarketplace/components/alert/macro.njk" import dmAlert %}
        {{ dmAlert({"text": "Alert body"}) }}
    """)
    template_output = template.render()

    assert template_output == snapshot
