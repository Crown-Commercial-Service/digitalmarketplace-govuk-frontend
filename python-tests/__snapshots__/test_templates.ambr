# name: test_dm_alert_with_no_attributes
  '
  
          
          <div
      data-module="dm-alert"
      role="alert"
      aria-label=""
      class="dm-alert"
      tabindex="-1"
  >
      <h2
          class="dm-alert__title dm-alert__title--large"
      >
          
      </h2>
      
          <div class="dm-alert__body">
              Alert body
          </div>
      
  </div>
      
  '
---
# name: test_template_renders[template_path0]
  '
  
    <footer class="govuk-footer " role="contentinfo">
    <div class="govuk-width-container ">
      
        <div class="govuk-footer__navigation">
          
            <div class="govuk-footer__section">
              <h2 class="govuk-footer__heading govuk-heading-m">About Digital Marketplace</h2>
              
                
                <ul class="govuk-footer__list ">
                  
                    
                      <li class="govuk-footer__list-item">
                        <a class="govuk-footer__link" href="#">
                          Contact
                        </a>
                      </li>
                    
                  
                    
                      <li class="govuk-footer__list-item">
                        <a class="govuk-footer__link" href="https://www.gov.uk/guidance/digital-marketplace-buyers-guide">
                          Services you can buy
                        </a>
                      </li>
                    
                  
                    
                      <li class="govuk-footer__list-item">
                        <a class="govuk-footer__link" href="https://www.gov.uk/guidance/digital-marketplace-suppliers-guide">
                          Services you can sell
                        </a>
                      </li>
                    
                  
                    
                      <li class="govuk-footer__list-item">
                        <a class="govuk-footer__link" href="https://www.gov.uk/government/organisations/crown-commercial-service">
                          About Crown Commercial Services
                        </a>
                      </li>
                    
                  
                    
                      <li class="govuk-footer__list-item">
                        <a class="govuk-footer__link" href="https://www.gov.uk/government/organisations/government-digital-service">
                          About Government Digital Services
                        </a>
                      </li>
                    
                  
                    
                      <li class="govuk-footer__list-item">
                        <a class="govuk-footer__link" href="#">
                          G-Cloud supplier A to Z
                        </a>
                      </li>
                    
                  
                </ul>
              
            </div>
          
            <div class="govuk-footer__section">
              <h2 class="govuk-footer__heading govuk-heading-m">Guidance for using Digital Marketplace</h2>
              
                
                <ul class="govuk-footer__list ">
                  
                    
                      <li class="govuk-footer__list-item">
                        <a class="govuk-footer__link" href="https://www.gov.uk/guidance/g-cloud-suppliers-guide">
                          Applying to sell on the G-Cloud framework
                        </a>
                      </li>
                    
                  
                    
                      <li class="govuk-footer__list-item">
                        <a class="govuk-footer__link" href="https://www.gov.uk/guidance/how-to-apply-to-sell-on-the-digital-outcomes-framework">
                          Applying to sell on the DOS framework
                        </a>
                      </li>
                    
                  
                    
                      <li class="govuk-footer__list-item">
                        <a class="govuk-footer__link" href="https://www.gov.uk/guidance/g-cloud-buyers-guide">
                          Buying on the G-Cloud framework
                        </a>
                      </li>
                    
                  
                    
                      <li class="govuk-footer__list-item">
                        <a class="govuk-footer__link" href="https://www.gov.uk/guidance/digital-outcomes-and-specialists-buyers-guide">
                          Buying on the DOS framework
                        </a>
                      </li>
                    
                  
                    
                      <li class="govuk-footer__list-item">
                        <a class="govuk-footer__link" href="https://www.gov.uk/guidance/the-crown-hosting-data-centres-framework-on-the-digital-marketplace">
                          Buying on the Crown Hosting framework
                        </a>
                      </li>
                    
                  
                </ul>
              
            </div>
          
        </div>
        <hr class="govuk-footer__section-break">
      
      <div class="govuk-footer__meta">
        <div class="govuk-footer__meta-item govuk-footer__meta-item--grow">
          
            <h2 class="govuk-visually-hidden">Support links</h2>
            
              <ul class="govuk-footer__inline-list">
                
                  <li class="govuk-footer__inline-list-item">
                    <a class="govuk-footer__link" href="#">
                      Terms and conditions
                    </a>
                  </li>
                
                  <li class="govuk-footer__inline-list-item">
                    <a class="govuk-footer__link" href="#">
                      Cookies
                    </a>
                  </li>
                
                  <li class="govuk-footer__inline-list-item">
                    <a class="govuk-footer__link" href="#">
                      Privacy notice
                    </a>
                  </li>
                
              </ul>
            
            
              <div class="govuk-footer__meta-custom">
                Built by the <a href="https://www.gov.uk/government/organisations/government-digital-service" class="govuk-footer__link">Government Digital Service</a>
              </div>
            
          
          <svg
            aria-hidden="true"
            focusable="false"
            class="govuk-footer__licence-logo"
            xmlns="http://www.w3.org/2000/svg"
            viewBox="0 0 483.2 195.7"
            height="17"
            width="41"
          >
            <path
              fill="currentColor"
              d="M421.5 142.8V.1l-50.7 32.3v161.1h112.4v-50.7zm-122.3-9.6A47.12 47.12 0 0 1 221 97.8c0-26 21.1-47.1 47.1-47.1 16.7 0 31.4 8.7 39.7 21.8l42.7-27.2A97.63 97.63 0 0 0 268.1 0c-36.5 0-68.3 20.1-85.1 49.7A98 98 0 0 0 97.8 0C43.9 0 0 43.9 0 97.8s43.9 97.8 97.8 97.8c36.5 0 68.3-20.1 85.1-49.7a97.76 97.76 0 0 0 149.6 25.4l19.4 22.2h3v-87.8h-80l24.3 27.5zM97.8 145c-26 0-47.1-21.1-47.1-47.1s21.1-47.1 47.1-47.1 47.2 21 47.2 47S123.8 145 97.8 145"
            />
          </svg>
          <span class="govuk-footer__licence-description">
            All content is available under the
            <a
              class="govuk-footer__link"
              href="https://www.nationalarchives.gov.uk/doc/open-government-licence/version/3/"
              rel="license"
            >Open Government Licence v3.0</a>, except where otherwise stated
          </span>
        </div>
        <div class="govuk-footer__meta-item">
          <a
            class="govuk-footer__link govuk-footer__copyright-logo"
            href="https://www.nationalarchives.gov.uk/information-management/re-using-public-sector-information/uk-government-licensing-framework/crown-copyright/"
          >Â© Crown copyright</a>
        </div>
      </div>
    </div>
  </footer>
  
  
  '
---
# name: test_template_renders[template_path10]
  '
  <section
      data-module="dm-banner"
      role="region"
      aria-label="<MockParam params.title>"
      class="dm-banner dm-banner--<MockParam params.type> <MockParam params.classes>"
  >
      <h<MockParam params.headingLevel>
          class="dm-banner__title dm-banner__title--large"
      >
          <MockParam params.title>
      </h<MockParam params.headingLevel>>
          <div class="dm-banner__body">
              <MockParam params.html>
          </div>
      
  </section>
  '
---
# name: test_template_renders[template_path11]
  ''
---
# name: test_template_renders[template_path1]
  ''
---
# name: test_template_renders[template_path2]
  '
  
  
  
  
  
  <div id="<MockParam params.id>" class="dm-cookie-banner govuk-width-container" data-module="dm-cookie-banner" role="region" aria-label="Cookie Banner" aria-describedby="dm-cookie-banner__heading">
    <div class="dm-cookie-banner__wrapper">
      <h2 class="dm-cookie-banner__heading govuk-heading-m" id="dm-cookie-banner__heading">
        Can we store analytics cookies on your device?
      </h2>
      <p class="govuk-body">
        Analytics cookies help us understand how our website is being used.
      </p>
      <div class="dm-cookie-banner__buttons">
        <button class="govuk-button dm-cookie-banner__button dm-cookie-banner__button--accept" type="submit" data-accept-cookies="true" aria-describedby="dm-cookie-banner__heading">
          Yes<span class="govuk-visually-hidden">, <MockParam params.serviceName> can store analytics cookies on your device</span>
        </button>
        <button class="govuk-button dm-cookie-banner__button dm-cookie-banner__button--reject" type="submit" data-accept-cookies="false" aria-describedby="dm-cookie-banner__heading">
          No<span class="govuk-visually-hidden">, <MockParam params.serviceName> cannot store analytics cookies on your device</span>
        </button>
        <a class="govuk-link dm-cookie-banner__link" href="<MockParam params.cookieInfoUrl>">How <MockParam params.serviceName> uses cookies</a>
      </div>
    </div>
  
    <div class="dm-cookie-banner__confirmation" tabindex="-1">
      <p class="dm-cookie-banner__confirmation-message govuk-body">
        You can <a class="govuk-link" href="<MockParam params.cookieSettingsUrl>">change your cookie settings</a> at any time.
      </p>
      <button class="dm-cookie-banner__hide-button govuk-link" data-hide-cookie-banner="true" role="link">Hide</button>
    </div>
  </div>
  '
---
# name: test_template_renders[template_path3]
  ''
---
# name: test_template_renders[template_path4]
  '
  <div
      data-module="dm-alert"
      role="alert"
      aria-label=""
      class="dm-alert dm-alert--<MockParam params.type> <MockParam params.classes>"
      tabindex="-1"
  >
      <h<MockParam params.headingLevel>
          class="dm-alert__title dm-alert__title--large"
      >
          <MockParam params.titleHtml>
      </h<MockParam params.headingLevel>>
      
          <div class="dm-alert__body">
              <MockParam params.html>
          </div>
      
  </div>
  '
---
# name: test_template_renders[template_path5]
  ''
---
# name: test_template_renders[template_path6]
  '
  <div class="dm-list-input params.classes  govuk-form-group--error">
      
    <fieldset class="govuk-fieldset <MockParam params.fieldset.classes>" aria-describedby="<MockParam params.fieldset.describedBy>" data-module="dm-list-input">
    
    <legend class="govuk-fieldset__legend <MockParam params.fieldset.legend.classes>">
    
      <h1 class="govuk-fieldset__heading">
        <MockParam params.fieldset.legend.text.__add__()>
      </h1>
    
    </legend>
    
  
      <div id="<MockParam params.id.__add__()>" class="govuk-hint <MockParam params.hint.classes.__radd__()>">
      <MockParam params.hint.html>
    </div>
    
      <span id="<MockParam params.id.__add__()>" class="govuk-error-message <MockParam params.errorMessage.classes.__radd__()>">
      <span class="govuk-visually-hidden"><MockParam params.errorMessage.visuallyHiddenText>:</span> <MockParam params.errorMessage.html>
    </span>
    
    
    
      <MockParam params.question_advice>
    
  
    <div class="dm-list-input__item-container" role="region" aria-live="assertive">
      
          
        <div class="dm-list-input__item dm-list-input__item--visible dm-list-input__item--error "><div class="dm-list-input__row">
            <div class="govuk-form-group govuk-form-group--error dm-list-input__form-group">
      <label class="govuk-label <MockParam params.items.__getitem__().label.classes.__radd__()>" for="<MockParam params.id>-1">
      <MockParam params.itemLabelPrefix.__radd__().__add__().__add__()>
    </label>
  
  
      
      
      <span id="<MockParam params.id>-1-error" class="govuk-error-message <MockParam params.items.__getitem__().errorMessage.classes.__radd__()>">
      <span class="govuk-visually-hidden"><MockParam params.items.__getitem__().errorMessage.visuallyHiddenText>:</span> <MockParam params.items.__getitem__().errorMessage.html>
    </span>
  
      <input class="govuk-input <MockParam params.items.__getitem__().classes.__radd__().__radd__()> govuk-input--error" id="<MockParam params.id>-1" name="<MockParam params.name>" type="<MockParam params.items.__getitem__().type>" value="<MockParam params.items.__getitem__().value>" aria-describedby="<MockParam params.items.__getitem__().describedBy.__add__().__add__()>" autocomplete="<MockParam params.items.__getitem__().autocomplete>" pattern="<MockParam params.items.__getitem__().pattern>" inputmode="<MockParam params.items.__getitem__().inputmode>">
    </div>
            
    
    
  
  
    
      
    
  
  <button type="button" class="govuk-button govuk-button--secondary dm-list-input__item-remove dm-list-input__item-remove--hidden" data-module="govuk-button">
    Remove <span class="govuk-visually-hidden">entry number <span class="dm-list-input__counter">1</span></span>
  </button>
  
  
          </div>
        </div>
      
    </div>
    
    
    
  
  
    
      
    
  
  <button type="button" class="govuk-button govuk-button--secondary dm-list-input__item-add dm-list-input__item-add--hidden" data-module="govuk-button">
    <MockParam params.addButtonName.__radd__().__add__()>
  </button>
  
  
    </fieldset>
  
  
  </div>
  '
---
# name: test_template_renders[template_path7]
  ''
---
# name: test_template_renders[template_path8]
  '
  
  
  
  
  
  
    
    
  
  
  
  
    
      
      
    
  
  
    <header class="govuk-header dm-admin-header" role="banner" data-module="govuk-header">
    <div class="govuk-header__container govuk-width-container">
      <div class="govuk-header__logo">
        <a href="/" class="govuk-header__link govuk-header__link--homepage">
          <span class="govuk-header__logotype">
            <svg
              aria-hidden="true"
              focusable="false"
              class="govuk-header__logotype-crown"
              xmlns="http://www.w3.org/2000/svg"
              viewBox="0 0 132 97"
              height="30"
              width="36"
            >
              <path
                fill="currentColor" fill-rule="evenodd"
                d="M25 30.2c3.5 1.5 7.7-.2 9.1-3.7 1.5-3.6-.2-7.8-3.9-9.2-3.6-1.4-7.6.3-9.1 3.9-1.4 3.5.3 7.5 3.9 9zM9 39.5c3.6 1.5 7.8-.2 9.2-3.7 1.5-3.6-.2-7.8-3.9-9.1-3.6-1.5-7.6.2-9.1 3.8-1.4 3.5.3 7.5 3.8 9zM4.4 57.2c3.5 1.5 7.7-.2 9.1-3.8 1.5-3.6-.2-7.7-3.9-9.1-3.5-1.5-7.6.3-9.1 3.8-1.4 3.5.3 7.6 3.9 9.1zm38.3-21.4c3.5 1.5 7.7-.2 9.1-3.8 1.5-3.6-.2-7.7-3.9-9.1-3.6-1.5-7.6.3-9.1 3.8-1.3 3.6.4 7.7 3.9 9.1zm64.4-5.6c-3.6 1.5-7.8-.2-9.1-3.7-1.5-3.6.2-7.8 3.8-9.2 3.6-1.4 7.7.3 9.2 3.9 1.3 3.5-.4 7.5-3.9 9zm15.9 9.3c-3.6 1.5-7.7-.2-9.1-3.7-1.5-3.6.2-7.8 3.7-9.1 3.6-1.5 7.7.2 9.2 3.8 1.5 3.5-.3 7.5-3.8 9zm4.7 17.7c-3.6 1.5-7.8-.2-9.2-3.8-1.5-3.6.2-7.7 3.9-9.1 3.6-1.5 7.7.3 9.2 3.8 1.3 3.5-.4 7.6-3.9 9.1zM89.3 35.8c-3.6 1.5-7.8-.2-9.2-3.8-1.4-3.6.2-7.7 3.9-9.1 3.6-1.5 7.7.3 9.2 3.8 1.4 3.6-.3 7.7-3.9 9.1zM69.7 17.7l8.9 4.7V9.3l-8.9 2.8c-.2-.3-.5-.6-.9-.9L72.4 0H59.6l3.5 11.2c-.3.3-.6.5-.9.9l-8.8-2.8v13.1l8.8-4.7c.3.3.6.7.9.9l-5 15.4v.1c-.2.8-.4 1.6-.4 2.4 0 4.1 3.1 7.5 7 8.1h.2c.3 0 .7.1 1 .1.4 0 .7 0 1-.1h.2c4-.6 7.1-4.1 7.1-8.1 0-.8-.1-1.7-.4-2.4V34l-5.1-15.4c.4-.2.7-.6 1-.9zM66 92.8c16.9 0 32.8 1.1 47.1 3.2 4-16.9 8.9-26.7 14-33.5l-9.6-3.4c1 4.9 1.1 7.2 0 10.2-1.5-1.4-3-4.3-4.2-8.7L108.6 76c2.8-2 5-3.2 7.5-3.3-4.4 9.4-10 11.9-13.6 11.2-4.3-.8-6.3-4.6-5.6-7.9 1-4.7 5.7-5.9 8-.5 4.3-8.7-3-11.4-7.6-8.8 7.1-7.2 7.9-13.5 2.1-21.1-8 6.1-8.1 12.3-4.5 20.8-4.7-5.4-12.1-2.5-9.5 6.2 3.4-5.2 7.9-2 7.2 3.1-.6 4.3-6.4 7.8-13.5 7.2-10.3-.9-10.9-8-11.2-13.8 2.5-.5 7.1 1.8 11 7.3L80.2 60c-4.1 4.4-8 5.3-12.3 5.4 1.4-4.4 8-11.6 8-11.6H55.5s6.4 7.2 7.9 11.6c-4.2-.1-8-1-12.3-5.4l1.4 16.4c3.9-5.5 8.5-7.7 10.9-7.3-.3 5.8-.9 12.8-11.1 13.8-7.2.6-12.9-2.9-13.5-7.2-.7-5 3.8-8.3 7.1-3.1 2.7-8.7-4.6-11.6-9.4-6.2 3.7-8.5 3.6-14.7-4.6-20.8-5.8 7.6-5 13.9 2.2 21.1-4.7-2.6-11.9.1-7.7 8.8 2.3-5.5 7.1-4.2 8.1.5.7 3.3-1.3 7.1-5.7 7.9-3.5.7-9-1.8-13.5-11.2 2.5.1 4.7 1.3 7.5 3.3l-4.7-15.4c-1.2 4.4-2.7 7.2-4.3 8.7-1.1-3-.9-5.3 0-10.2l-9.5 3.4c5 6.9 9.9 16.7 14 33.5 14.8-2.1 30.8-3.2 47.7-3.2z"
              ></path>
              <image src="/assets/images/govuk-logotype-crown.png" xlink:href="" class="govuk-header__logotype-crown-fallback-image" width="36" height="32"></image>
            </svg>
            <span class="govuk-header__logotype-text">
              GOV.UK
            </span>
          </span>
          
          <span class="govuk-header__product-name">
            Digital Marketplace Admin
          </span>
          
        </a>
      </div>
      
      <div class="govuk-header__content">
      
      
      <button type="button" class="govuk-header__menu-button govuk-js-header-toggle" aria-controls="navigation" aria-label="Show or hide Top Level Navigation">Menu</button>
      <nav>
        <ul id="navigation" class="govuk-header__navigation " aria-label="Top Level Navigation">
          
            
              <li class="govuk-header__navigation-item">
                <a class="govuk-header__link" href="https://www.gov.uk/guidance/buying-and-selling-on-the-digital-marketplace">
                  Guidance
                </a>
              </li>
            
          
            
              <li class="govuk-header__navigation-item">
                <a class="govuk-header__link" href="#">
                  Help
                </a>
              </li>
            
          
            
              <li class="govuk-header__navigation-item">
                <a class="govuk-header__link" href="#">
                  Log out
                </a>
              </li>
            
          
        </ul>
      </nav>
      
      </div>
      
    </div>
  </header>
  
  
  '
---
# name: test_template_renders[template_path9]
  ''
---
